### PR TITLE
feat: implement walk-forward backtesting and model performance dashboard

### LINKED ISSUE #66: [FEATURE] Historical backtest dashboard: validate model accuracy and calibration on past IPL seasons
Hi! Was looking through `application.py` to understand how the model
gets evaluated, and noticed there's no built-in way to see how
accurate the XGBoost predictor actually is on real matches. The
README pitches "Fintech-grade Cricket Intelligence," and the first
thing any fintech demand of a predictive model is a backtest on
historical data. Wanted to flag this as a gap and propose adding
the missing surface.

### Problem / Motivation

The current `train_model()` function fits XGBoost on ball-by-ball
data from `matches.csv` and `deliveries.csv` and serves win
probabilities live. But there's nothing in the app today that
answers the questions a careful user would ask:

- How often is the model right when it predicts >70% confidence?
- Is a "60% win" prediction actually right 60% of the time, or is
  the model overconfident / underconfident?
- Does accuracy degrade in the death overs (16-20), or against
  specific venues?
- Has any recent change to the pipeline made the model better or
  worse?

Without those answers, a user staring at a "73% win probability"
has no way to know whether 73% means anything. This is the
"backtest your model" hygiene that the Fintech-grade tagline
implies but the app doesn't currently show.

Issue #16 proposes adding 5-fold cross-validation accuracy to the
Analysis page, which would surface a single number ("82% accurate").
This proposal is complementary, not overlapping: it covers the
*visual* side (calibration plot, phase breakdown, drift over time)
that a single accuracy number can't show.

### Proposed Solution

A new Streamlit page **"Model Performance"** in the sidebar,
alongside the existing Dashboard and Match Analysis tabs.

**1. Walk-forward train/test split by season.** Train on IPL
2008-2020, hold out 2021+ as the test set. This avoids the
information leakage that an i.i.d. random split would cause and
matches the post-COVID era split that cricket fans recognize.

**2. Replay backtest.** For each held-out match, replay it
ball-by-ball using the model trained on prior seasons only. Record
predicted probability at each delivery.

**3. Three core visualizations:**

- **Calibration plot.** Bucket predictions into 10 deciles
  (0-10%, 10-20%, ..., 90-100%). For each bucket, plot predicted
  probability against actual win rate. A well-calibrated model
  hugs the diagonal. This is the single most useful diagnostic for
  a probability model.
- **Brier score over time.** Lower is better. Plot per-season to
  spot drift as cricket itself evolved.
- **Accuracy by match phase.** Powerplay (1-6), middle (7-15),
  death (16-20). Models often do well in the middle and badly at
  the death; the chart exposes that.

**4. Summary metrics card** with overall accuracy at the 50%
threshold, Brier score, log loss, and the sample size (matches +
balls evaluated).

**5. Caching.** Wrap the backtest pipeline in `@st.cache_data` so
the page loads instantly on revisit and only recomputes when the
underlying model or data changes.

```python
@st.cache_data
def run_backtest(test_seasons: list[int]) -> pd.DataFrame:
    """
    Returns one row per delivery in the test set with columns:
    match_id, over, ball, predicted_prob, actual_outcome,
    season, phase ('powerplay' | 'middle' | 'death')
    """
    ...
```

### Alternatives considered

**One-shot accuracy on a random train-test split.** Simpler but
leaks information across seasons (the model gets to learn from
2023 matches and then "predict" 2019 ones). Walk-forward by
season is the standard for time-series-flavored data and matches
how the model would actually be used.

**5-fold cross-validation only (issue #16).** That covers the
"single accuracy number" use case. This issue covers the visual
diagnostics (calibration, phase, drift). Both can coexist; if #16
lands first, the metric card on this page can also surface its
number alongside the new charts.

**A standalone Jupyter notebook in the repo.** Reaches contributors
but not the demo's actual users. The point of a visible page is to
let any user see for themselves that the model is good, which is
much more persuasive than telling them.

**scikit-learn `learning_curve` / `validation_curve`.** Those
measure training dynamics during fitting, not deployment
performance. Useful for tuning, not for answering "is this model
good in production."

### Acceptance Criteria

- New Streamlit page "Model Performance" reachable from the sidebar
  alongside Dashboard and Match Analysis.
- Backtest pipeline trains on IPL 2008-2020 and evaluates on 2021+
  with no leakage.
- Calibration plot, Brier-score-over-time chart, and accuracy-by-
  phase chart all render.
- Summary metric card shows overall accuracy at 50% threshold,
  Brier score, log loss, and sample size.
- Page loads in under 2 seconds after the first compute, via
  `@st.cache_data`.
- No regression in the existing Dashboard or Match Analysis pages.
- A short note in the README explains what the new page shows and
  what the calibration plot means.

Two design questions worth confirming before I start:

1. Train/test split as a hardcoded 2008-2020 / 2021+ for v1, or
   configurable via a sidebar slider? My instinct is hardcoded for
   v1 to keep the page focused; a slider can be a follow-up.
2. Cached backtest results in `@st.cache_data` only, or also
   persisted to disk as `backtest_results.parquet`? Disk caching
   survives restarts but adds a file to the repo or `.gitignore`.
   My instinct is in-memory only for v1.

### PR DESCRIPTION
**Pull request by manandev18**
## Description
Closes #66 
This PR addresses the lack of visible model evaluation metrics by introducing a dedicated **Model Performance** page. It implements a walk-forward backtesting pipeline to evaluate the XGBoost prediction engine on unseen holdout seasons, proving the "fintech-grade" reliability of the model.

## Changes Included
- **Added `run_backtest()` pipeline**: Implemented a walk-forward train/test split (Trains on IPL 2008-2017, Tests on IPL 2018-2019) to prevent time-series data leakage.
- **New Sidebar Navigation**: Added a `⚡ Model Performance` tab alongside Dashboard and Match Analysis.
- **Summary Metrics Card**: Surfaces overall Accuracy (50% threshold), Brier Score, Log Loss, and testing sample size.
- **Diagnostic Visualizations (Plotly)**:
  - **Calibration Plot**: Maps predicted win probability against actual win rates across 10 decile bins to ensure the model hugs the diagonal.
  - **Brier Score Over Time**: Tracks model drift across holdout seasons.
  - **Accuracy by Match Phase**: Breaks down prediction accuracy across Powerplay, Middle, and Death overs.
- **Caching**: Wrapped the backtest pipeline in `@st.cache_data` for instant page load on revisit.
- **README Update**: Added the walk-forward backtesting features to the Prediction Engine feature list.

## Files Modified
- `application.py` (Added backtest logic, UI rendering, and routing)
- `README.md` (Updated features list)

## Screenshots
image
image
image

## Verification
- Validated that `city` and other required features map correctly during the evaluation merge.
- Renamed the native `matches.csv` string `result` column internally to `match_result_type` to prevent collision with the binary ML win label. 
- All visualizations render successfully with no regressions to existing pages.

### CODE REVIEW

**Comment by manandev18** [comment]
Hi @Arnav-Singh-5080, please check the Pull request and please give the gssoc:approved and additional tags if needed

**Comment by manandev18** [comment]
Hi @Arnav-Singh-5080 , please review the pull request

**Comment by Arnav-Singh-5080** [comment]
Great

**Comment by manandev18** [comment]
> Great

Please label the issue as GSSoc: approved with additional labels

**Comment by manandev18** [comment]
@Arnav-Singh-5080  Please label the issue as gssoc:approved with additional labels.

**Comment by manandev18** [comment]
@Arnav-Singh-5080 
Screenshot_20260523-162034

**Comment by manandev18** [comment]
@Arnav-Singh-5080 , please label the pr as gssoc:approved
